unmarshaller.go

```diff
@@ -1,117 +1,120 @@
-package gocsv
+package commando
 
 import (
-	"encoding/csv"
 	"fmt"
+	"io"
 	"reflect"
 )
 
 // Unmarshaller is a CSV to struct unmarshaller.
 type Unmarshaller struct {
-	reader                 *csv.Reader
-	Headers                []string
-	fieldInfoMap           []*fieldInfo
-	MismatchedHeaders      []string
-	MismatchedStructFields []string
-	outType                reflect.Type
+	config *validConfig
+	line   int
+	reader Reader
 }
 
-// NewUnmarshaller creates an unmarshaller from a csv.Reader and a struct.
-func NewUnmarshaller(reader *csv.Reader, out interface{}) (*Unmarshaller, error) {
+// NewUnmarshaller is a convenience function which allocates and
+// returns a new Unmarshaller.
+func NewUnmarshaller(holder interface{}, reader Reader) (*Unmarshaller, error) {
+	return (&Config{Holder: holder}).NewUnmarshaller(reader)
+}
+
+// NewUnmarshaller creates an unmarshaller from a Reader and a struct.
+func (c *Config) NewUnmarshaller(reader Reader) (*Unmarshaller, error) {
 	headers, err := reader.Read()
 	if err != nil {
 		return nil, err
 	}
-	headers = normalizeHeaders(headers)
 
-	um := &Unmarshaller{reader: reader, outType: reflect.TypeOf(out)}
-	err = validate(um, out, headers)
+	vc, err := c.validate(headers)
 	if err != nil {
 		return nil, err
 	}
+
+	um := &Unmarshaller{
+		reader: reader,
+		config: vc,
+		line:   1,
+	}
+
 	return um, nil
 }
 
-// Read returns an interface{} whose runtime type is the same as the struct that
-// was used to create the Unmarshaller.
+// Read returns an interface{} whose runtime type is the same as the
+// struct that was used to create the Unmarshaller.
 func (um *Unmarshaller) Read() (interface{}, error) {
-	row, err := um.reader.Read()
+	value, _, err := um.ReadUnmatched()
+	return value, err
+}
+
+// ReadAllCallback calls cb for every record Read() from um.
+func ReadAllCallback(um *Unmarshaller, onSuccess func(interface{}) error, onError func(error) error) error {
+	for {
+		rec, err := um.Read()
+		if err == io.EOF {
+			return nil
+		} else if err != nil {
+			if handlerErr := onError(err); handlerErr != nil {
+				return handlerErr
+			}
+		}
+
+		if err := onSuccess(rec); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// wrapLine wraps err, including the line the error occurred on.
+func wrapLine(err error, line int) error {
 	if err != nil {
-		return nil, err
+		return fmt.Errorf("On line %d: %w", line, err)
```

Comment:
[Code review comments](https://github.com/golang/go/wiki/CodeReviewComments#error-strings) would suggest lowercasing, here.

```suggestion
		return fmt.Errorf("on line %d: %w", line, err)
```

Alternatively, we could consider implementing our own error type here, which might let us do more interesting things in the `onError` callbacks.